src/messaging.rs

```diff
@@ -167,4 +181,28 @@ impl<NodeUid: Ord> NetworkInfo<NodeUid> {
     pub fn num_faulty(&self) -> usize {
         self.num_faulty
     }
+
+    pub fn secret_key(&self) -> &SecretKey<Bls12> {
+        &self.secret_key
+    }
+
+    pub fn public_key_set(&self) -> &PublicKeySet<Bls12> {
+        &self.public_key_set
+    }
+
+    /// The canonical numbering of all nodes.
+    ///
+    /// FIXME: To avoid multiple computations of the same result, caching should be introduced.
```

Comment:
If we do indeed need this at all, I'd vote for computing it only once, in the constructor.

Reply:
I'm trying to find out if the following modification of `NetworkInfo` is workable:
```
pub struct NetworkInfo<'a, NodeUid: 'a> {
    our_uid: &'a NodeUid,
    all_uids: BTreeSet<&'a NodeUid>,
    num_nodes: usize,
    num_faulty: usize,
    secret_key: SecretKey<Bls12>,
    public_key_set: PublicKeySet<Bls12>,
    /// An ordered list of node IDs for indexing purposes.
    node_list: Vec<&'a NodeUid>,
}
```
Here, `node_list` is computed in the constructor. After a straightforward refactoring, the main library compiles. However tests and examples would require some restructuring because many functions rely on passing node IDs by value. Pass-by-reference definitely works in the library but becomes quite complicated outside it.

Reply:
I think it's okay to require `NodeUid: Clone` everywhere, to remove the lifetime parameter.

But a `Vec<NodeUid>` still isn't optimal for index lookup (linear time). A `HashMap<NodeUid, usize>` would find the index in constant time.